tests/apollo/test_skvbc_checkpoints.py

```diff
@@ -375,6 +397,146 @@ async def test_checkpoint_propagation_after_f_nodes_including_primary_isolated(s
+
+        with log.start_action(action_type='Expected_failure_when_filling_and_waiting_for_checkpoint'):
+            with self.assertRaises(trio.TooSlowError):
+                  await skvbc.fill_and_wait_for_checkpoint(
+                    initial_nodes=bft_network.all_replicas(without=byzantine_replica_ids),
+                    num_of_checkpoints_to_add=5,
+                    verify_checkpoint_persistency=False
+                    )
+
+        key1 = ['checkpoint_msg', 'Counters', 'number_of_checkpoint_mismatch']
+        key2 = ['replica', 'Gauges', 'lastExecutedSeqNum']
+        for replica_id in bft_network.all_replicas() :
+            last_stored_cp = await bft_network.wait_for_checkpoint(replica_id)
+            last_executed_seq_num = await bft_network.metrics.get(replica_id, *key2)
+            number_of_checkpoint_mismatch = await bft_network.metrics.get(replica_id, *key1)
+            assert last_stored_cp == 5, \
+                f"Replica {replica_id} last_stored_cp={last_stored_cp} != 5"
+            assert last_executed_seq_num >= 750, \
+                f"Replica {replica_id} last_executed_seq_num={last_executed_seq_num} < 750"
+            if replica_id not in byzantine_replica_ids:
+                n = len(byzantine_replica_ids) * 5
+                assert number_of_checkpoint_mismatch == n, \
+                    f"Replica {replica_id} number_of_checkpoint_mismatch={number_of_checkpoint_mismatch} != {n}"
+
+    @with_trio
+    @with_bft_network(start_replica_cmd_with_corrupted_checkpoint_msgs(corrupt_checkpoints_from_replica_ids={ 0 }))
+    async def test_checkpoint_propagation_after_corrupting_checkpoint_msg_for_primary(self, bft_network):
+        """
+        This test verifies that the primary reaches the same checkpoint number as the one in the `bft_network`,
+        even though it sends incorrect data in its checkpoint messages.
+
+        1) Start all replicas in the given `bft_network`
+        2) Get the current primary and verify the assumption that it is 0
+        3) Send enough requests to trigger 4 checkpoints
+        4) Verify that all replicas reached checkpoint 4, including the primary.
+        5) Verify that all honest replicas received 4 mismatched checkpoint messages.
+        """
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+
+        current_primary = await bft_network.get_current_primary()
+        assert current_primary == 0, "Unexpected initial primary."
+
+        await skvbc.fill_and_wait_for_checkpoint(
+            initial_nodes=bft_network.all_replicas(without={current_primary}),
+            num_of_checkpoints_to_add=4,
+            verify_checkpoint_persistency=False
+        )
+
+        key1 = ['checkpoint_msg', 'Counters', 'number_of_checkpoint_mismatch']
+        key2 = ['replica', 'Gauges', 'lastExecutedSeqNum']
+        for replica_id in bft_network.all_replicas(without={current_primary}) :
```

Comment:
The current primary has to be included in here in order to verify -> 4) Verify that all replicas reached checkpoint 4, including the primary.

Reply:
I'm going to fix things differently:

- fill_and_wait_for_checkpoint will wait to al 7 backup replicas
- In the loop verify mismatches for not primary, and minimal  seqno for all